### ISSUE TITLE
Error in weight normalization code?

### ISSUE THREAD

**Issue by johnpjust** [nicola-decao/BNAF#1]
In the BNAF-->maskedweights-->get_weights function, there is a line as follows...

`w = torch.exp(self._weight) * self.mask_d + self._weight * self.mask_o`

I believe the `torch.exp(self._weight) * self.mask_d` should be `torch.exp(self._diag_weight) * self.mask_d` though right?

**Comment by nicola-decao** [comment]
Hi, I think the line is correct. Probably I used an odd name for the variable but `self._diag_weight` is a vector not a matrix of the norms for the weight normalization. That is why it is multiplied after a couple of lines. The line `w = torch.exp(self._weight) * self.mask_d + self._weight * self.mask_o` computes the unnormalized weight matrix and then I divide by its norm and I multiply by `self._diag_weight`. See https://github.com/nicola-decao/BNAF/blob/1a629a2c6159909cc7c365d4092595782b003698/bnaf.py#L195

**Comment by johnpjust** [comment]
Agreed that it is indeed a vector and dimension-wise what i said wouldn't work, but where (prior to that line) do you actually set the diagonals of the weight matrix equal to the values actually contained in self._diag_weight so that your unnormalized weight matrix has the correct diagonal values?  You initialize the diagonal weights separately in the masked_weight function (using a random_uniform distribution) from the rest of the off-diagonal weights contained in self._weight.

I may be missing something somewhere, but seems like the `w_squared_norm` will be wrong if you don't set the diagonals in self._weight to self._diag_weights somewhere...?

I don't program in pytorch, but in Tensorflow it should look something like this...

`w = tf.multiply(tf.exp(tf.diag(self._diag_weight)), self.mask_d) + tf.multiply(self._weight, self.mask_o)`

And now the `w_squared_norm `will be correct.
